engine/src/main/java/org/pentaho/di/pan/PanCommandExecutor.java

```diff
@@ -0,0 +1,491 @@
+
+    } else if ( !Utils.isEmpty( exportRepo ) ) {
+      // Export the repository
+      System.out.println( BaseMessages.getString( getPkgClazz(), "Pan.Log.ExportingObjectsRepToFile", "" + exportRepo ) );
+      repository.getExporter().exportAllObjects( null, exportRepo, directory, "all" );
+      System.out.println( BaseMessages.getString( getPkgClazz(), "Pan.Log.FinishedExportObjectsRepToFile", "" + exportRepo ) );
+    }
+  }
+
+  public Trans loadTransFromRepository( Repository repository, String dirName, String transName ) throws Exception {
+
+    if ( Utils.isEmpty( transName ) ) {
+      System.out.println( BaseMessages.getString( getPkgClazz(), "Pan.Error.NoTransNameSupplied" ) );
+      return null;
+    }
+
+    RepositoryDirectoryInterface directory = loadRepositoryDirectory( repository, dirName, "Pan.Error.NoRepProvided",
+            "Pan.Log.Allocate&ConnectRep", "Pan.Error.CanNotFindSpecifiedDirectory" );
+
+    if ( directory == null ) {
+      return null; // not much we can do here
+    }
+
+    // Add the IMetaStore of the repository to our delegation
+    if ( repository.getMetaStore() != null && getMetaStore() != null ) {
+      getMetaStore().addMetaStore( repository.getMetaStore() );
+    }
+
+    logDebug( "Pan.Log.LoadTransInfo" );
+    TransMeta transMeta = repository.loadTransformation( transName, directory, null, true, null );
+
+    logDebug( "Pan.Log.AllocateTrans" );
+    Trans trans = new Trans( transMeta );
+    trans.setRepository( repository );
+    trans.setMetaStore( getMetaStore() );
+
+    return trans; // return transformation loaded from the repo
+  }
+
+  public Trans loadTransFromFilesystem( String initialDir, String filename, String jarFilename, Serializable base64Zip ) throws Exception {
+
+    Trans trans = null;
+
+    File zip;
+    if ( base64Zip != null && ( zip = decodeBase64ToZipFile( base64Zip, true ) ) != null ) {
+      // update filename to a meaningful, 'ETL-file-within-zip' syntax
+      filename = "zip:file:" + File.separator + File.separator + zip.getAbsolutePath() + "!" + filename;
+    }
+
+    // Try to load the transformation from file
+    if ( !Utils.isEmpty( filename ) ) {
+
+      String filepath = filename;
+      // If the filename starts with scheme like zip:, then isAbsolute() will return false even though the
+      // the path following the zip is absolute. Check for isAbsolute only if the fileName does not start with scheme
+      if ( !KettleVFS.startsWithScheme( filename ) && !FileUtil.isFullyQualified( filename ) ) {
+        filepath = initialDir + filename;
+      }
+
+      logDebug( "Pan.Log.LoadingTransXML", "" + filepath );
+      TransMeta transMeta = new TransMeta( filepath );
+      trans = new Trans( transMeta );
+    }
+
+    if ( !Utils.isEmpty( jarFilename ) ) {
+
+      try {
+
+        logDebug( "Pan.Log.LoadingTransJar", jarFilename );
+
+        InputStream inputStream = PanCommandExecutor.class.getResourceAsStream( jarFilename );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *inputStream*. The resource is referenced at line 383. The resource is closed at line 386. There are other execution paths that do not contain closure statements, for example, when InputStream.read throws an exception. Ensure *inputStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).